src/ray/object_manager/pull_manager.cc

```diff
@@ -193,66 +243,76 @@ void PullManager::UpdatePullsBasedOnAvailableMemory(size_t num_bytes_available)
+    RAY_LOG(ERROR) << "Margin required " << margin_required;
+    DeactivateUntilMarginAvailable("task args request", task_argument_bundles_,
+                                   /*retain_min=*/0, /*quota_margin=*/margin_required,
+                                   &highest_task_req_id_being_pulled_,
+                                   &object_ids_to_cancel);
 
     // Activate the next wait request if we have space.
-    if (num_bytes_being_pulled_ < num_bytes_available_) {
-      wait_requests_remaining = ActivateNextPullBundleRequest(
-          wait_request_bundles_, &highest_wait_req_id_being_pulled_, &objects_to_pull);
-    } else {
-      break;
-    }
+    wait_requests_remaining = ActivateNextPullBundleRequest(
+        wait_request_bundles_, &highest_wait_req_id_being_pulled_,
+        /*respect_quota=*/true, &objects_to_pull);
   }
 
   // Do the same but for task arg requests (lowest priority).
   // allowed for task arg requests.
-  while (num_bytes_being_pulled_ < num_bytes_available_) {
-    if (!ActivateNextPullBundleRequest(task_argument_bundles_,
+  while (ActivateNextPullBundleRequest(task_argument_bundles_,
                                        &highest_task_req_id_being_pulled_,
-                                       &objects_to_pull)) {
-      break;
-    }
+                                       /*respect_quota=*/true, &objects_to_pull)) {
   }
 
   // While we are over capacity, deactivate requests starting from the back of the queues.
-  DeactivateUntilWithinQuota("task args request", task_argument_bundles_,
-                             &highest_task_req_id_being_pulled_, &object_ids_to_cancel);
-  DeactivateUntilWithinQuota("wait request", wait_request_bundles_,
-                             &highest_wait_req_id_being_pulled_, &object_ids_to_cancel);
+  DeactivateUntilMarginAvailable(
+      "task args request", task_argument_bundles_, min_active_pulls_, /*quota_margin=*/0L,
+      &highest_task_req_id_being_pulled_, &object_ids_to_cancel);
+  DeactivateUntilMarginAvailable("wait request", wait_request_bundles_, min_active_pulls_,
+                                 /*quota_margin=*/0L, &highest_wait_req_id_being_pulled_,
+                                 &object_ids_to_cancel);
   // It should always be possible to stay under the available memory by
   // canceling all requests.
   if (!RayConfig::instance().plasma_unlimited()) {
-    DeactivateUntilWithinQuota("get request", get_request_bundles_,
-                               &highest_get_req_id_being_pulled_, &object_ids_to_cancel);
-    RAY_CHECK(num_bytes_being_pulled_ <= num_bytes_available_);
+    DeactivateUntilMarginAvailable("get request", get_request_bundles_, min_active_pulls_,
+                                   /*quota_margin=*/0L, &highest_get_req_id_being_pulled_,
+                                   &object_ids_to_cancel);
+    RAY_CHECK(!OverQuota() || num_active_bundles_ <= min_active_pulls_) << DebugString();
```

Comment:
```suggestion
    RAY_CHECK(!OverQuota() || num_active_bundles_ >= min_active_pulls_) << DebugString();
```
I think?

Reply:
It's actually correct-- if over quota, then it must be due to the min active bundles, so check it's <= that.